app/frontend/src/utils/createAssociation.ts

```diff
@@ -1,76 +1,108 @@
 import { AssociationProps } from "./Props";
 
 class AssociationElement {
-    public op: AssociationProps;
+    public assProps: AssociationProps;
 
-    constructor(options: AssociationProps) {
-        this.op = options;
+    constructor(props: AssociationProps, margin: number) {
+        this.assProps = props;
     }
 
-    draw(ctx: CanvasRenderingContext2D, lineWidth: number) {
-        ctx.save();
-        ctx.lineWidth = lineWidth;
-        ctx.strokeStyle = "black";
+    draw(ctx: CanvasRenderingContext2D, margin: number, lineWidth: number) {
 
-        // 處理不同的 assType
-        switch (this.op.assType) {
-            case 0: // 普通實線
-                ctx.setLineDash([]);
-                break;
-            case 1: // 虛線
-                ctx.setLineDash([5, 5]);
-                break;
-            case 2: // 繼承（箭頭）
-                ctx.setLineDash([]);
-                this.drawArrow(ctx);
-                break;
-            // 可擴充其他關係如聚合、依賴等
-            default:
-                ctx.setLineDash([]);
+        if (this.assProps.deltaX === 0 && this.assProps.deltaY === 0) {
+            this.drawNormalAss(ctx, margin, lineWidth);
+        }
+        else{
+            this.drawSelfAss(ctx, margin, lineWidth);
         }
 
-        // 畫主線
-        ctx.beginPath();
-        ctx.moveTo(this.op.startX, this.op.startY);
-        ctx.lineTo(this.op.endX, this.op.endY);
-        ctx.stroke();
-        ctx.setLineDash([]);
+        this.drawArrow(ctx, margin, lineWidth);
 
-        // 畫屬性文字
-        this.op.attributes.forEach(attr => {
-            const midX = this.op.startX + (this.op.endX - this.op.startX) * attr.ratio;
-            const midY = this.op.startY + (this.op.endY - this.op.startY) * attr.ratio;
+        // Draw label text(s) if present
+        this.drawText(ctx, margin);
+    }
 
-            ctx.font = `${attr.fontSize}px ${attr.fontFile}`;
-            ctx.fillStyle = "black";
-            ctx.fillText(attr.content, midX + 5, midY - 5); // 偏移避免覆蓋線條
-        });
+    drawText(ctx: CanvasRenderingContext2D, margin: number) {
+        // attributes: { content, fontSize, fontStyle, fontFile, ratio }[]
+        if (Array.isArray(this.assProps.attributes)) {
+            this.assProps.attributes.forEach(attr => {
+                if (attr && typeof attr.content === "string") {

```

Comment:
```suggestion
                if (attr) {

```

Redundant 'typeof' check: 'content' always has type 'string'

Reply:
solved